project4/name_server/bootstrap.cpp

```diff
@@ -1,40 +1,200 @@
+  } else if (request.mutable_predecessor_info()->IsInitialized() || request.mutable_successor_info()->IsInitialized()) {
+    // this should never happen.
+    // error
+    return;
+  }
+  client_ = std::make_unique<message_passing::Client>(threadpool_, entering_nameserver_);
+  if (client_->Send(request) < 0 ) {
+    // error
+  }
+}
 
-void Bootstrap::Insert(int key, std::string val) {}
+void Bootstrap::LookUp(uint key) {
+  // key-val resides in this nameserver
+  auto itr = pairs_.find(key);
+  if (itr != pairs_.end()) {
+    // key-val pair found , now print
+    console_task_->SendConsole(std::string("Value: ").append(itr->second));
+    console_task_->SendConsole("Contacted: 0");
+  } else if (successor_ == bootstrap_) {
+    console_task_->SendConsole("Key not found.");
+  } else {
+    consistent_hash_msgs::LookUpResult lookup;
+    lookup.set_id(0);
+    lookup.set_key(key);
+    client_ =
+        std::make_unique<message_passing::Client>(threadpool_, successor_);
+    if (client_->Send(lookup) < 0) {
+      // error
+    }
+  }
+}
 
-void Bootstrap::Delete(int key, std::string val) {}
+void Bootstrap::Insert(uint key, std::string val) {
+  if (key >= bounds_.first && key <= bounds_.second) {
+    // key-val should be inserted in this bootstrap
+    std::pair<int, std::string> pair(key, val);
+    pairs_.insert(pair);
+    console_task_->SendConsole("Inserted into: 0");
+    console_task_->SendConsole("Contacted: 0");
+  } else if (successor_ == bootstrap_) {
+    console_task_->SendConsole("Key out of range.");
+  } else {
+    consistent_hash_msgs::InsertResult insert;
+    insert.set_id(0);
+    insert.set_key(0);
+    client_ =
+        std::make_unique<message_passing::Client>(threadpool_, successor_);
+    if (client_->Send(insert) < 0) {
+      // error
+    }
+  }
+}
 
-void Bootstrap::HandleRequest(
-    const consistent_hash_msgs::EntranceInformation& request) {}
+void Bootstrap::Delete(uint key) {
+  // key-val resides in this nameserver
+  auto itr = pairs_.find(key);
+  if (itr != pairs_.end()) {
+    // key-val pair found , now delete
+    pairs_.erase(itr);
+    console_task_->SendConsole("Contacted: 0");
+  } else if (successor_ == bootstrap_) {
+    console_task_->SendConsole("Key not found.");
+  } else {
+    consistent_hash_msgs::DeleteResult delete_r;
+    delete_r.set_delete_success(false);
+    delete_r.set_key(key);
+    client_ =
+        std::make_unique<message_passing::Client>(threadpool_, successor_);
+    if (client_->Send(delete_r) < 0) {
+      // error
+    }
+  }
+}
 
-void Bootstrap::HandleRequest(
-    const consistent_hash_msgs::ExitInformation& request) {}
+void Bootstrap::HandleRequest(consistent_hash_msgs::LookUpResult& request) {
+  // denote this server as contacted
+  auto ids = request.server_ids();
+  request.set_server_ids(ids.size(), 0);
 
-void Bootstrap::HandleRequest(
-    const consistent_hash_msgs::LookUpResult& request) {}
+  if (request.value().empty()) {
```

Comment:
Yeah, I'm not exactly sure. I do know it popped up on the CLion autocomplete so I figured I'd give it a try. It would make sense for it to have an `empty()`, kinda like how messages have the `IsInitialized`

Reply:
Ah, the reason this works is because `value` is an `std::string`, so this is not going to do exactly what you want it to. However, if you assume that no use will ever set a value to an empty string, then it will work, but I'm not sure we can assume that. You might need to add a separate `is_found` value to the message.